UPGRADE-GUIDE.md

```diff
@@ -0,0 +1,159 @@
+
+### `rack.hijack?` now only applies to partial hijack
+
+Previously, both full and partial hijiack were controlled by the presence and value of `rack.hijack?`. Now, it only applies to partial hijack (which itself has been effectively replaced by streaming bodies).
+
+### `rack.hijack` alone indicates that you can execute a full hijack
+
+Previously, `rack.hijack?` had to be truthy, as well as having `rack.hijack` present in the request environment. Now, the presence of the `rack.hijack` callback is enough.
+
+### `rack.hijack_io` is removed
+
+Previously, the server would try to set `rack.hijack_io` into the request environment when `rack.hijack` was invoked for a full hijack. This was often impossible if a middleware had called `env.dup`, so this requirement has been dropped entirely.
+
+### `rack.input` is no longer required to be rewindable
+
+Previosuly, `rack.input` was required to be rewindable, i.e. `io.seek(0)` but this was only generally possible with a file based backing, which prevented efficient streaming of request bodies. Now, `rack.input` is not required to be rewindable.
+
+## Response Changes
+
+### Response must be mutable
+
+Rack 3 requires the response Array `[status, headers, body]` to be mutable. Existing code that uses a frozen response will need to be changed:
+
+```ruby
+NOT_FOUND = [404, {}, ["Not Found"]].freeze
+
+def call(env)
+  ...
+	return NOT_FOUND
+end
+```
+
+should be rewritten as:
+
+```ruby
+def not_found
+	[404, {}, ["Not Found"]]
+end
+
+def call(env)
+  ...
+	return not_found
+end
+```
+
+Note there is a subtle bug in the former version: the headers hash is mutable and can be modified, and these modifications can leak into subsequent requests.
+
+### Response headers must be a mutable hash
+
+Rack 3 requires response headers to be a mutable hash. Previously it could be any object that would respond to `#each` and yield `key`/`value` pairs. Previously, the following was acceptable:
+
+```ruby
+def call(env)
+  return [200, [['content-type', 'text/plain']], ["Hello World]]
+end
+```
+
+Now you must use a hash instance:
+
+```ruby
+def call(env)
+  return [200, {'content-type' => 'text/plain'}, ["Hello World]]
+end
+```
+
+This ensures middleware can predictably update headers as needed.
+
+### Response Headers must be lower case
+
+Rack 3 requires all response headers to be lower case. This is to simplify fetching and updating response headers. Previously you had to use something like `Rack::HeadersHash`
+
+```ruby
+def call(env)
+	response = @app.call(env)
+	# HeaderHash must allocate internal objects and compute lower case keys:
+	headers = Rack::Utils::HeaderHash[response[1]]
+
+	cache_response(headers['ETag'], response)
+
+	...
+end
+```
+
+but now you must just use the normal form for HTTP header:
+
+```ruby
+def call(env)
+	response = @app.call(env)
+	# A plain hash with lower case keys:
+	headers = response[1]
+
+	cache_response(headers['etag'], response)
+
+	...
+end
+```
```

Comment:
We should mention `Rack::Headers` here, and how it can be used to make apps designed to work with Rack 2 usable on Rack 3 with minimal changes.